File: CSharp/PatternsInAction/Patterns In Action/BusinessObjects/BusinessRules/ValidateRange.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BusinessObjects.BusinessRules
{
    /// <summary>
    /// Validates a range (min and max) for a given data type.
    /// </summary>
    public class ValidateRange : BusinessRule
    {
        private ValidationDataType DataType { get; set; }
        private ValidationOperator Operator { get; set; }

        private object Min { get; set; }
        private object Max { get; set; }

        public ValidateRange(string propertyName, object min, object max,
            ValidationOperator @operator, ValidationDataType dataType)
            : base(propertyName)
        {
            Min = min;
            Max = max;

            Operator = @operator;
            DataType = dataType;

            ErrorMessage = propertyName + " must be between " + Min + " and " + Max;
        }

        public ValidateRange(string propertyName, string errorMessage, object min, object max,
            ValidationOperator @operator, ValidationDataType dataType)
            : this (propertyName,  min,  max,@operator, dataType)
        {
            ErrorMessage = errorMessage;
        }

        public override bool Validate(BusinessObject businessObject)
        {
            try
            {
                string value = GetPropertyValue(businessObject).ToString();

                switch (DataType)
                {
                    case ValidationDataType.Integer:

                        int imin = int.Parse(Min.ToString());
                        int imax = int.Parse(Max.ToString());
                        int ival = int.Parse(value);

                        return (ival >= imin && ival <= imax);

                    case ValidationDataType.Double:
                        double dmin = double.Parse(Min.ToString());
                        double dmax = double.Parse(Max.ToString());
                        double dval = double.Parse(value);

                        return (dval >= dmin && dval <= dmax);

                    case ValidationDataType.Decimal:
                        decimal cmin = decimal.Parse(Min.ToString());
                        decimal cmax = decimal.Parse(Max.ToString());
                        decimal cval = decimal.Parse(value);

                        return (cval >= cmin && cval <= cmax);

                    case ValidationDataType.Date:
                        DateTime tmin = DateTime.Parse(Min.ToString());
                        DateTime tmax = DateTime.Parse(Max.ToString());
                        DateTime tval = DateTime.Parse(value);

                        return (tval >= tmin && tval <= tmax);

                    case ValidationDataType.String:

                        string smin = Min.ToString();
                        string smax = Max.ToString();

                        int result1 = string.Compare(smin, value);
                        int result2 = string.Compare(value, smax);

                        return result1 >= 0 && result2 <= 0;
                }
                return false;
            }
            catch { return false; }
        }
    }
}

File: VB/PatternsInAction/Patterns In Action/BusinessObjects/BusinessRules/ValidateRange.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace BusinessObjects.BusinessRules
	''' <summary>
	''' Validates a range (min and max) for a given data type.
	''' </summary>
	Public Class ValidateRange
		Inherits BusinessRule
        Private _dataType As ValidationDataType
        Private Property DataType() As ValidationDataType
            Get
                Return _dataType
            End Get
            Set(ByVal value As ValidationDataType)
                _dataType = value
            End Set
        End Property
        Private _operator As ValidationOperator
        Private Property TheOperator() As ValidationOperator
            Get
                Return _operator
            End Get
            Set(ByVal value As ValidationOperator)
                _operator = value
            End Set
        End Property

        Private _min As Object
        Private Property Min() As Object
            Get
                Return _min
            End Get
            Set(ByVal value As Object)
                _min = value
            End Set
        End Property
        Private _max As Object
        Private Property Max() As Object
            Get
                Return _max
            End Get
            Set(ByVal value As Object)
                _max = value
            End Set
        End Property

        Public Sub New(ByVal propertyName As String, ByVal min As Object, ByVal max As Object, ByVal theOperator As ValidationOperator, ByVal dataType As ValidationDataType)
            MyBase.New(propertyName)
            Me.Min = min
            Me.Max = max

            Me.TheOperator = theOperator
            Me.DataType = dataType

            ErrorMessage = propertyName & " must be between " & CStr(Me.Min) & " and " & CStr(Me.Max)
        End Sub

        Public Sub New(ByVal propertyName As String, ByVal errorMessage As String, ByVal min As Object, ByVal max As Object, ByVal theOperator As ValidationOperator, ByVal dataType As ValidationDataType)
            Me.New(propertyName, min, max, theOperator, dataType)
            Me.ErrorMessage = errorMessage
        End Sub

		Public Overrides Function Validate(ByVal businessObject As BusinessObject) As Boolean
			Try
				Dim value As String = GetPropertyValue(businessObject).ToString()

				Select Case DataType
					Case ValidationDataType.Integer

						Dim imin As Integer = Integer.Parse(Min.ToString())
						Dim imax As Integer = Integer.Parse(Max.ToString())
						Dim ival As Integer = Integer.Parse(value)

						Return (ival >= imin AndAlso ival <= imax)

					Case ValidationDataType.Double
						Dim dmin As Double = Double.Parse(Min.ToString())
						Dim dmax As Double = Double.Parse(Max.ToString())
						Dim dval As Double = Double.Parse(value)

						Return (dval >= dmin AndAlso dval <= dmax)

					Case ValidationDataType.Decimal
						Dim cmin As Decimal = Decimal.Parse(Min.ToString())
						Dim cmax As Decimal = Decimal.Parse(Max.ToString())
						Dim cval As Decimal = Decimal.Parse(value)

						Return (cval >= cmin AndAlso cval <= cmax)

					Case ValidationDataType.Date
						Dim tmin As DateTime = DateTime.Parse(Min.ToString())
						Dim tmax As DateTime = DateTime.Parse(Max.ToString())
						Dim tval As DateTime = DateTime.Parse(value)

						Return (tval >= tmin AndAlso tval <= tmax)

					Case ValidationDataType.String

						Dim smin As String = Min.ToString()
						Dim smax As String = Max.ToString()

						Dim result1 As Integer = String.Compare(smin, value)
						Dim result2 As Integer = String.Compare(value, smax)

						Return result1 >= 0 AndAlso result2 <= 0
				End Select
				Return False
			Catch
				Return False
			End Try
		End Function
	End Class
End Namespace
